### PR TITLE
cpp: Add changes to operator new from C++17 and C++20

### PR DESCRIPTION
**Pull request by ppryga-nordic**
C++ subsystem defines exception throwing operator new.
Though the implementation never throws an exception.

If used by an application it should verify if requested
memory was allocated. This is against C++ standard.
If an application does not support exceptions or does not
want to call throwing new operator, it should use a
specialization of a new operator that makes sure it
never throws bad_alloc exception. The cpp subsystem
does not provide this specialization.

C++17 delivered another specialization of a new operator,
that allows to allocate a memory with requested alignment.
Implementation of this change requires an extension to 
minimal libc. There is no allocator that allows for request 
particular alignment.

C++20 standard added [[nodiscard]] attribute to operator new
definitions. In case value returned by operator new is ignored,
compiler emits a warning.

### CODE REVIEW

**Comment by ppryga-nordic** [comment]
Please have a look on the PR, I'm welcome for comments about this change/extension proposal.

Since exceptions are supported by C++ subsystem, at least there is a Kconfig option that enables it, maybe it would be good 
to make the throwing operator new to throw `std::bad_alloc` as it is expected by C++ specification. What do you think?

**Comment by aborisovich** [comment]
I think application users should decide for themselves - so operator `new` should `throw std::bad_alloc` as default behavior implies. From Zephyr perspective I think we may provide code with [terminate handler](https://en.cppreference.com/w/cpp/error/terminate_handler) using [set_terminate](https://en.cppreference.com/w/cpp/error/set_terminate) function that would just do some while(true) looping (I know Xtensa toolchain implementation does this as exception handling part) as optional but enabled by default setting.

If we talk about compiling with `-fno-exceptions` we can add some alternative new operator overloads that return error codes.
We can probably do detection on whether exceptions are enabled or not using Kconfig option as you suggested. I think it would be too much of a pain to do detection using compiler attributes or features since we use so many toolchains and every of them has their own syntax...

**Comment by stephanosio** [comment]
> 1. Add Kconfig option that is used to determine whether C++ uses exceptions or not.

There already exists `CONFIG_EXCEPTIONS`:

https://github.com/zephyrproject-rtos/zephyr/blob/691175f026a6cc9661ef398417a05b220be9d308/subsys/cpp/Kconfig#L65-L69

Also note that C++ exception support depends on the `CONFIG_LIB_CPLUSPLUS` (full C++ standard library implementation such as libstdc++) because the Zephyr C++ subsystem (i.e. the minimal C++ support library implementation) lacks the infrastructure to properly support exceptions, so it is currently not feasible to throw exceptions inside Zephyr C++ subsystem's `new` operator implementation.

We could expand the scope of the Zephyr C++ subsystem to support exceptions so that exceptions can be used without requiring a full C++ standard library implementation; but, that requires the a lot of compiler/ABI-specific handling and increases toolchain dependency which we are trying to avoid.

In general, the rule of thumb for now is that, if you want to use C++ exceptions, use a full C++ standard library (`CONFIG_LIB_CPLUSPLUS=y`) and not Zephyr's minimal C++ subsystem.

File: subsys/cpp/cpp_new.cpp
Comment (stephanosio) [inline subsys/cpp/cpp_new.cpp]: Nit: spurious space

```diff
-void* operator new[](std::size_t size, std::align_val_t al )
+void* operator new[](std::size_t size, std::align_val_t al)
```

File: subsys/cpp/cpp_new.cpp
Comment (stephanosio) [inline subsys/cpp/cpp_new.cpp]: Nit: spurious space

```diff
-void* operator new (std::size_t size, std::align_val_t al,
+void* operator new(std::size_t size, std::align_val_t al,
```

**Review by stephanosio — COMMENTED** [review]
No objection to the general direction of the changes here, just some nits.

File: lib/libc/minimal/source/stdlib/malloc.c
Comment (ppryga-nordic) [reply lib/libc/minimal/source/stdlib/malloc.c]: This was left from my testing code. I didn't spot that I left it here.
Should be removed now.

**Comment by ppryga-nordic** [comment]
> I think application users should decide for themselves - so operator `new` should `throw std::bad_alloc` as standard implies. From Zephyr perspective I think we may provide code with [terminate handler](https://en.cppreference.com/w/cpp/error/terminate_handler) using [set_terminate](https://en.cppreference.com/w/cpp/error/set_terminate) function that would just do some while(true) looping (I know Xtensa toolchain implementation does this as exception handling part) as optional but enabled by default setting.
> 
> If we talk about compiling with `-fno-exceptions` we can add some alternative new operator overloads that return error codes. We can probably do detection on whether exceptions are enabled or not using Kconfig option as you suggested. I think it would be too much of a pain to do detection using compiler attributes or features since we use so many toolchains and every of them has their own syntax...
> 
> Thought about CMake - [target_compile_features](https://cmake.org/cmake/help/v3.1/command/target_compile_features.html) for a while to detect exception handling at CMake level but unfortunately list of compilers supported by CMake is limited and I know that Xtensa compiler is not on the list (as probably other embedded toolchains).
> 
> We can even think about writing [custom allocator](https://en.cppreference.com/w/cpp/memory/allocator) for exception-less applications to use, however it does look like a quite hard subject (at least for me, never wrote one).
> 
> As far as I know, when exception is thrown and you compile with `-fno-exceptions` function [std::abort()](https://en.cppreference.com/w/cpp/utility/program/abort) is called that causes abnormal program termination **unless**:
> 
> > Causes abnormal program termination unless [SIGABRT](https://en.cppreference.com/w/cpp/utility/program/SIG_types) is being caught by a signal handler passed to [std::signal](https://en.cppreference.com/w/cpp/utility/program/signal) and the handler does not return.
> 
> From this docs it looks like we can register a signal handler that would be called every time the exception is thrown and `-fno-exceptions` flag is used. It would prevent sudden program crash and every function throwing would just return [SIGN_ERR](https://en.cppreference.com/w/cpp/utility/program/SIG_ERR) on error.
> 
> ### To sum up:
> What I would do for now:
> 
> 1. Add Kconfig option that is used to determine whether C++ uses exceptions or not.
> 2. If Kconfig option is enabled and exceptions are used, we should write and register simple implementation of exception handler that just does `while(true) { };`
> 3. If Kconfig option is disabled and exceptions are not used, we should write and register simple implementation of signal handler that would not crash program on throw but return SIG_ERR. This can be used to implement "return code style new" operator and I think it's more "native C++" way to do this than just reinventing sets of return code logic.
> 4. We can think about this custom allocator but my C++ skills are probably not good enough for this ;-)

In the reference to the comment:
1. The change to `minimal libc` is blocked by https://github.com/zephyrproject-rtos/zephyr/issues/30105. The `C11` is not yet enabled hence the code will not compile.
2. The change to already available `operator new` (to make it throwing `std::bad_alloc` exception), in my humble opinion, should got through API change process. I would deliver that as a separate PR hence it is a bigger change because exceptions are allowed and enabled by `CONFIG_EXCEPTIONS` but there are not standard exceptions available in C++ subsystem. I'd like to look into the topic a bit deeper to get some feeling how complex would it be to implement some minimal handling of exceptions in `minimal C++ subsystem`.

>

**Comment by stephanosio** [comment]
> 2\. I'd like to look into the topic a bit deeper to get some feeling how complex would it be to implement some minimal handling of exceptions in `minimal C++ subsystem`.

The biggest problem is that, as mentioned above, C++ language support implementation is highly compiler type, version and ABI (hence, also architecture)-specific-- implementing C++ exception support would require the Zephyr C++ subsystem (it should really be called "Zephyr minimal C++ support library) to be closely tied to the toolchain.

For instance, even if we add C++ exception handling support for GCC, it will only work with GCC and supporting other compilers will require further work. That is one of the reasons why the C++ standard library development usually goes alongside the compiler development (e.g. libstdc++, which is part of the GCC project) and it is distributed as part of a toolchain.

If we really wanted to add exception handling support to the Zephyr C++ subsystem for GCC, we basically need to re-implement our own libsupc++ (see https://github.com/gcc-mirror/gcc/tree/master/libstdc%2B%2B-v3/libsupc%2B%2B).

**Comment by stephanosio** [comment]
Also I would suggest reading up on the following issues:

* https://github.com/zephyrproject-rtos/zephyr/issues/45785
* https://github.com/zephyrproject-rtos/zephyr/issues/31281
* The "To be" portion of https://github.com/zephyrproject-rtos/zephyr/issues/36644#issue-933805956

**Comment by ppryga-nordic** [comment]
@stephanosio how many toolchains do we support now? I'm wondering about amount of work required to provide exceptions to `minimal C++ subsystem`. Making it dependent on selected number of toolchains like gcc, clang does not sound like right way.

**Comment by stephanosio** [comment]
> @stephanosio how many toolchains do we support now? I'm wondering about amount of work required to provide exceptions to `minimal C++ subsystem`. Making it dependent on selected number of toolchains like gcc, clang does not sound like right way.

https://github.com/zephyrproject-rtos/zephyr/tree/main/cmake/compiler

There is also IAR support work in progress.

If you can explain what your goals and requirements are (and why you cannot use `libstdc++`), then I could try revising the future C++ support plans to reflect that (e.g. adding a separate more footprint-optimised libstdc++ variant in the Zephyr SDK).

**Comment by ppryga-nordic** [comment]
Well I'd like to have an `operator new` that works according to standard, I mean throws. I'm not sure if I can use `libstdc++` due to memory constraints. That I need to verify. Because of that I'm also checking other roads like extension of a `minimal C++ subsystem` with exceptions.

**Comment by stephanosio** [comment]
> Well I'd like to have an `operator new` that works according to standard, I mean throws. I'm not sure if I can use `libstdc++` due to memory constraints. That I need to verify.

FYI, the footprint requirement for libstdc++ has been significantly reduced in Zephyr SDK 0.15.0 -- see https://github.com/zephyrproject-rtos/sdk-ng/issues/527#issuecomment-1216975682.

**Comment by ppryga-nordic** [comment]
> > Well I'd like to have an `operator new` that works according to standard, I mean throws. I'm not sure if I can use `libstdc++` due to memory constraints. That I need to verify.
> 
> FYI, the footprint requirement for libstdc++ has been significantly reduced in Zephyr SDK 0.15.0 -- see [zephyrproject-rtos/sdk-ng#527 (comment)](https://github.com/zephyrproject-rtos/sdk-ng/issues/527#issuecomment-1216975682).

Thank you for information. I'll make the evaluation.

**Review by cfriedt — DISMISSED** [review]

**Comment by ppryga-nordic** [comment]
Updated code to address not resolved comments.

**Review by stephanosio — APPROVED** [review]

**Review by aborisovich — APPROVED** [review]

**Comment by fabiobaltieri** [comment]
(tagged for v3.3.0, feel free to change if this should be considered a fix for v3.2.0 instead)

**Pull request merged**
